streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -388,76 +392,107 @@ public void process(final ProcessorSupplier<K, V> processorSupplier, String... s
             Serde<V1> otherValueSerde,
             boolean outer) {
 
-        Set<String> allSourceNodes = ensureJoinableWith((AbstractStream<K>) other);
-
-        StateStoreSupplier thisWindow = Stores.create(windows.name() + "-this")
-                .withKeys(keySerde)
-                .withValues(thisValueSerde)
-                .persistent()
-                .windowed(windows.maintainMs(), windows.segments, true)
-                .build();
-
-        StateStoreSupplier otherWindow = Stores.create(windows.name() + "-other")
-                .withKeys(keySerde)
-                .withValues(otherValueSerde)
-                .persistent()
-                .windowed(windows.maintainMs(), windows.segments, true)
-                .build();
+        return doJoin(other,
+                      joiner,
+                      windows,
+                      keySerde,
+                      thisValueSerde,
+                      otherValueSerde,
+                      new DefaultJoin(outer));
+    }
+
+    private <V1, R> KStream<K, R> doJoin(KStream<K, V1> other,
+                                         ValueJoiner<V, V1, R> joiner,
+                                         JoinWindows windows,
+                                         Serde<K> keySerde,
+                                         Serde<V> thisValueSerde,
+                                         Serde<V1> otherValueSerde,
+                                         KStreamImplJoin join) {
+        KStreamImpl<K, V> joinThis = this;
+        KStreamImpl<K, V1> joinOther = (KStreamImpl) other;
+
+        if (joinThis.repartitionRequired) {
+            joinThis = joinThis.repartitionForJoin(keySerde, thisValueSerde, joinOther
+                .repartitionRequired);
+        }
 
-        KStreamJoinWindow<K, V> thisWindowedStream = new KStreamJoinWindow<>(thisWindow.name(), windows.before + windows.after + 1, windows.maintainMs());
-        KStreamJoinWindow<K, V1> otherWindowedStream = new KStreamJoinWindow<>(otherWindow.name(), windows.before + windows.after + 1, windows.maintainMs());
+        if (joinOther.repartitionRequired) {
+            joinOther = joinOther.repartitionForJoin(keySerde, otherValueSerde, this
+                .repartitionRequired);
+        }
 
-        KStreamKStreamJoin<K, R, V, V1> joinThis = new KStreamKStreamJoin<>(otherWindow.name(), windows.before, windows.after, joiner, outer);
-        KStreamKStreamJoin<K, R, V1, V> joinOther = new KStreamKStreamJoin<>(thisWindow.name(), windows.before, windows.after, reverseJoiner(joiner), outer);
+        // if both input streams need repartitioning we need to make sure
+        // that they are co-partitioned as this means the repartitioned streams
+        // will also be co-partitioned
```

Comment:
Thanks for the detailed response @guozhangwang. It was done like this to ensure that the two re-partitioned streams end up with the same partition count, however, you are correct, in this case it is really only the re-partitioned streams that need to be co-partitioned, not the original streams.
To make this work i believe I'll need to an extra method to the TopologyBuilder to link the internal re-partition topics to the sourceTopics. We'll need this information to work out the number of partitions for the internal topics. Something like:
`TopologyBuilder copartitionInternalTopics(final Set<String> sourceNodes,
                                                     final Set<String> repartitionNodes)`
 
This method would map each of the repartitionNodes topics to a set of sourceNode topics. Then during `ensureCopartitioning` in `StreamPartitionAssignor` we can use this info to find the correct partition count for these topics, i.e., max partition count of the source topics